### PR TITLE
[4.0] RangeInfo: Add a new range info kind that describes part of a parent expression. rdar://32039874 

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 1]
```diff
@@ -746,6 +777,17 @@ struct RangeResolver::Implementation {
     return true;
   }
 
+  bool nodeContainSelection(ASTNode Node) {
+    // If the selection starts before the node, return false.
+    if (SM.isBeforeInBuffer(Start, Node.getStartLoc()))
+      return false;
+    // If the node ends before the selection, return false.
```
Comment (benlangmuir) [inline lib/IDE/SwiftSourceDocInfo.cpp]: Nitpick: these comments don't add much beyond what's in the source code; I'd just remove them.

**Review by benlangmuir — APPROVED** [review]

**Comment by nkcsgexi** [comment]
Approved by Ted via email

**Pull request merged**